src/Sulu/Component/Content/Document/Syncronization/DocumentRegistrator.php

```diff
@@ -0,0 +1,214 @@
+<?php
+
+/*
+ * This file is part of Sulu.
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Component\Content\Document\Syncronization;
+
+use PHPCR\Util\PathHelper;
+use Sulu\Component\Content\Document\Behavior\SynchronizeBehavior;
+use Sulu\Component\DocumentManager\Behavior\Mapping\ParentBehavior;
+use Sulu\Component\DocumentManager\DocumentManagerInterface;
+
+/**
+ * Class responsible for registering a document from the default document manager
+ * with the publish document manager including any documents dependent on
+ * the document to be synchronized.
+ */
+class DocumentRegistrator
+{
+    /**
+     * @var DocumentManagerInterface
+     */
+    private $defaultManager;
+
+    /**
+     * @var DocumentManagerInterface
+     */
+    private $publishManager;
+
+    public function __construct(
+        DocumentManagerInterface $defaultManager,
+        DocumentManagerInterface $publishManager
+    ) {
+        $this->defaultManager = $defaultManager;
+        $this->publishManager = $publishManager;
+    }
+
+    /**
+     * Register the incoming DDM document with any existing PHPCR node in the
+     * PDM.
+     *
+     * If the PDM already has the incoming PHPCR node then we need to register
+     * the existing PHPCR node from the PDM PHPCR session with the incoming DDM
+     * document (otherwise the system will attempt to create a new document and
+     * fail).
+     *
+     * We also ensure that any immediately related documents are also registered.
+     *
+     * @param SynchronizeBehavior $document
+     */
+    public function registerDocumentWithPDM(SynchronizeBehavior $document)
+    {
+        $this->registerSingleDocumentWithPDM($document);
+
+        $metadata = $this->defaultManager->getMetadataFactory()->getMetadataForClass(get_class($document));
+        // iterate over the field mappings for the document, if they resolve to
+        // an object then try and register it with the PDM.
+        foreach (array_keys($metadata->getFieldMappings()) as $field) {
```

Comment:
So the document might be registered multiple times? In which case does it make sense to have multiple metadata for one document?